src/Integration/LocalServices/ErrorListInfoBarController.cs

```diff
@@ -462,6 +491,190 @@ private ProjectViewModel FindProject(Uri serverUri, string projectKey)
+                }
+
+                this.host = host;
+                this.TokenSource = new CancellationTokenSource();
+            }
+
+            internal /*for testing purposes*/ CancellationTokenSource TokenSource
+            {
+                get;
+            }
+
+            internal /*for testing purpose*/ System.Threading.Tasks.Task BackgroundTask
+            {
+                get;
+                private set;
+            }
+
+            public void QueueCheckIfUpdateIsRequired(Action updateAction)
+            {
+                if (updateAction == null)
+                {
+                    throw new ArgumentNullException(nameof(updateAction));
+                }
+
+                Debug.Assert(this.BackgroundTask == null, "Not expecting this method to be called more than once");
+
+                var projectSystem = this.host.GetService<IProjectSystemHelper>();
+                projectSystem.AssertLocalServiceIsNotNull();
+
+                IEnumerable<LanguageGroup> languages = projectSystem.GetFilteredSolutionProjects()
+                    .Select(p => LanguageGroupHelper.GetProjectGroup(p))
+                    .Distinct();
+
+                if(!languages.Any())
+                {
+                    return;
+                }
+
+                var solutionBinding = this.host.GetService<ISolutionBindingSerializer>();
+                solutionBinding.AssertLocalServiceIsNotNull();
+
+                var binding = solutionBinding.ReadSolutionBinding();
+                if (binding == null)
+                {
+                    return;
+                }
+
+                if (binding.Profiles == null || binding.Profiles.Count == 0)
+                {
+                    // Old binding, force refresh immediately
+                    VsShellUtils.WriteToGeneralOutputPane(this.host, Strings.SonarLintProfileCheckNoProfiles);
+                    updateAction();
+                    return;
+                }
+
+                VsShellUtils.WriteToGeneralOutputPane(this.host, Strings.SonarLintProfileCheck);
+
+                CancellationToken token = this.TokenSource.Token;
+                this.BackgroundTask = System.Threading.Tasks.Task.Run(() =>
+                {
+                    if (this.ProcessQualityProfileChanges(binding, languages, token))
+                    {
+                        this.host.UIDispatcher.BeginInvoke(new Action(()=>
+                        {
+                            // We mustn't update if was cancelled
+                            if(!token.IsCancellationRequested)
+                            {
+                                updateAction();
+                            }
+                        }));
+                    }
+                }, token);
+            }
+
+            private bool ProcessQualityProfileChanges(BoundSonarQubeProject binding, IEnumerable<LanguageGroup> projectLanguageGroups, CancellationToken token)
```

Comment:
Documentation: what does the return value indicate - whether an update is required or not?